### PR TITLE
MULTI or DUAL EndStop config fails to compile 

### PR DESCRIPTION
**Pull request by descipher**
### Description

MULTI or DUAL endstops need to allow MAX or MIN based plug usage. 

### Requirements

Two or more physical end stops on any AXIS 

### Benefits

Allows users to use a second plus such as XMAX on a MIN base config etc.

### Configurations

```cpp
#define USE_ZMIN_PLUG
//Enabled for dual endstop
#define USE_XMAX_PLUG   // used as the second z stepper end limit switch

#define Z_MULTI_ENDSTOPS
#if ENABLED(Z_MULTI_ENDSTOPS)
  #define Z2_USE_ENDSTOP          _XMAX_
```

### Related Issues

#22774
#23411

### CODE REVIEW

**Comment by thinkyhead** [comment]
Looking over the endstop logic, there is a lot of convolution that could possibly be simplified. A general overhaul of endstop assignment and logic is probably about due. Anyway, the error in the case of issue 23411 was a misconfiguration. The Z homing direction was towards min, when it should have been towards max.

**Comment by thinkyhead** [comment]
I've attempted to get rid of the `USE_*_PLUG` options because they're annoying. Instead, the pins defined for the board are allowed to persist throughout, but various options are checked to decide whether to use them. It's important for the symbol `X_MAX_PIN` to exist, for example, if `Z2_USE_ENDSTOP` is `_XMAX_`, but we don't want to initialize it as both `X_MAX_PIN` and as `Z2_MIN_PIN`. Hopefully, the changes I've applied here will simplify configuration, reduce the build size, clarify the endstops logic, and still result in proper behavior in all cases.

I've left open the possibility for future adjustments to allow setting `*_HOME_DIR` to 0 if no homing is wanted on an axis, but I have not added any logic to `G28` to handle that case yet.

**Comment by ellensp** [comment]
can I add a related request to this endstop re-rewrite?

Can we change #define Z_MIN_PROBE_USES_Z_MIN_ENDSTOP_PIN to
#define PROBE_IS_PLUGGED_INTO PROBE_PORT // Z_MIN_PORT or PROBE_PORT 
or something similar. (in the same way we have #define SDCARD_CONNECTION LCD or ONBOARD)

two reasons for this...   apparently Z_MIN_PROBE_USES_Z_MIN_ENDSTOP_PIN is not clear enough. 
secondly to deliberately break all the old install tutorial that people are blindly following that incorrectly says to set Z_MIN_PROBE_USES_Z_MIN_ENDSTOP_PIN always

**Comment by thinkyhead** [comment]
> can I add a related request to this endstop re-rewrite?

no

**Comment by descipher** [comment]
Read through it, I will flash a printer and put it through various checks. Including config breakage injection .. lol

**Comment by thinkyhead** [comment]
Ambitiously, I'm trying to simplify the probe / endstop configuration and handling, but haven't wrangled everything yet. I need to look closer to the case where Z homes to min and the probe shares the Z min endstop, both when the probe is used for homing and when it's not.

The basic idea is to detect when `Z_MIN_PROBE_PIN` == `Z_MIN_PIN` and let that replace (or set as a conditional) `Z_MIN_PROBE_USES_Z_MIN_ENDSTOP_PIN`. That single change would be interesting to try by itself apart from the rest of this incomplete rework.

At any rate, it looks like the `USE_*_PLUG` options may be relatively easy to drop.

**Comment by descipher** [comment]
Yes, not easy to visualize, certainly takes time to see all the states.

**Comment by thinkyhead** [comment]
I'm going to sleep on it and come back with a fresh perspective … and a flow chart.

**Comment by descipher** [comment]
> I'm going to sleep on it and come back with a fresh perspective … and a flow chart.

Sleep is always good for clarity.

**Comment by descipher** [comment]
Let me know when to run tests.
Did an initial test .. Z probe fails to be sensed
#define USE_XMIN_PLUG
#define USE_YMIN_PLUG
#define USE_ZMIN_PLUG
#define X_HOME_DIR -1
#define Y_HOME_DIR -1
#define Z_HOME_DIR -1
#define X_MIN_ENDSTOP_INVERTING true // Set to true to invert the logic of the endstop.
#define Y_MIN_ENDSTOP_INVERTING true // Set to true to invert the logic of the endstop.
#define Z_MIN_ENDSTOP_INVERTING false // Set to true to invert the logic of the endstop.
#define FIX_MOUNTED_PROBE
#define Z_MIN_PROBE_ENDSTOP_INVERTING false // Set to true to invert the logic of the probe.
#define Z_MIN_PROBE_USES_Z_MIN_ENDSTOP_PIN

**Comment by descipher** [comment]
Turning on PINS_DEBUGING ..

**Comment by thinkyhead** [comment]
This is just a very crude attempt to brute force a successful compile, and I don't expect anything here to work. I'm documenting endstops in detail and I'll be able to apply some elements of this, but certainly not all.

**Comment by descipher** [comment]
This should be helpful:
```
10:48:06.908 > PIN:   2   Port: E4        X_MAX_PIN                              protected
10:48:06.912 > PIN:   3   Port: E5        X_MIN_PIN                              protected 
10:48:06.915 > .                          X_STOP_PIN                             protected
10:48:07.006 > PIN:  14   Port: J1        Y_MIN_PIN                              protected
10:48:07.014 > .                          Y_STOP_PIN                             protected
10:48:07.014 > PIN:  15   Port: J0        Y_MAX_PIN                              protected 
10:48:07.021 > PIN:  18   Port: D3        Z_MIN_PIN                              protected
10:48:07.024 > .                          Z_STOP_PIN                             protected
10:48:07.028 > .                          Z_MIN_PROBE_PIN                        protected
10:48:07.030 > PIN:  19   Port: D2        Z_MAX_PIN                              protected
```
Edit: Cleaned

File: Marlin/src/inc/Conditionals_post.h [hunk 1]
```diff
@@ -2304,83 +2332,77 @@
...
-#if _HAS_STOP(Y,MIN)
+#if   Y_HOME_TO_MIN && _HAS_STOP(Y,MIN)
   #define HAS_Y_MIN 1
-#endif
-#if _HAS_STOP(Y,MAX)
+#elif Y_HOME_TO_MAX && _HAS_STOP(Y,MAX)
   #define HAS_Y_MAX 1
 #endif
-#if _HAS_STOP(Z,MIN)
+#if   Z_HOME_TO_MIN && _HAS_STOP(Z,MIN) && Z_MIN_PROBE_PIN != Z_MIN_PIN
```
Comment (descipher) [inline Marlin/src/inc/Conditionals_post.h]:
```
#if   Z_HOME_TO_MIN && _HAS_STOP(Z,MIN) Z_MIN_PROBE_PIN != Z_MIN_PIN
```
Results in no HAS_Z_MIN def
```
// *** CI Compatibility with old configs ***
#if ENABLED(Z_MIN_PROBE_USES_Z_MIN_ENDSTOP_PIN)
  #undef Z_MIN_PROBE_PIN
  #define Z_MIN_PROBE_PIN Z_MIN_PIN
#endif
```

**Comment by descipher** [comment]
Found additional issues when testing configuration options.
```
 #if HAS_Z_AXIS
    if (stepper.axis_is_moving(Z_AXIS)) {
      if (stepper.motor_direction(Z_AXIS_HEAD)) { // Z -direction. Gantry down, bed up.

        #if HAS_Z_MIN && !USE_PROBE_FOR_Z_HOMING
          PROCESS_ENDSTOP_Z(MIN);
```
The configuration option USE_PROBE_FOR_Z_HOMING
is not a bool and when defined compiling fails.

File: Marlin/src/inc/Conditionals_post.h (hunk 1 above)
Comment (thinkyhead) [reply Marlin/src/inc/Conditionals_post.h]: I expect much if this is wrong. I am documenting endstops and will have changes when I have changes.

**Pull request merged**